server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -365,24 +368,73 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
+            } else {
+                XContentHelper.merge(mergedRawMapping, rawMapping, ((parent, key, oldValue, newValue) -> {
+                    switch (key) {
+                        case "type" -> {
+                            // todo: verify this check is valid
+                            if (oldValue.equals(newValue) == false
+                                && (OBJECT_FIELD_TYPES.contains(String.valueOf(oldValue))
+                                    || OBJECT_FIELD_TYPES.contains(String.valueOf(newValue)))) {
+                                throw new MapperParsingException(
+                                    "can't merge a non object mapping [" + parent + "] with an object mapping"
+                                );
+                            }
+                        }
+                        case "subobjects" -> {
+                            if (oldValue.equals(newValue) == false) {
+                                throw new MapperParsingException("contradicting subobjects settings provided for field: " + parent);
+                            }
+                        }
+                        case "required" -> {
+                            if ("_routing".equals(parent) && oldValue != newValue) {
+                                throw new MapperParsingException("contradicting `_routing.required` settings");
+                            }
+                        }
+                    }
+                    return newValue;
+                }));
+            }
+        }
+        return (mergedRawMapping != null) ? doMerge(type, reason, mergedRawMapping) : null;
+    }
+
     public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
         final DocumentMapper currentMapper = this.mapper;
         if (currentMapper != null && currentMapper.mappingSource().equals(mappingSource)) {
             return currentMapper;
         }
-        synchronized (this) {
-            Mapping incomingMapping = parseMapping(type, mappingSource);
-            Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason);
-            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
-            // We should identify these cases reliably and save expensive serialization here
-            DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
-            if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-                return newMapper;
-            }
-            this.mapper = newMapper;
-            assert assertSerialization(newMapper);
+        Map<String, Object> mappingSourceAsMap = MappingParser.convertToMap(mappingSource);
+        return doMerge(type, reason, mappingSourceAsMap);
+    }
+
+    private synchronized DocumentMapper doMerge(String type, MergeReason reason, Map<String, Object> mappingSourceAsMap) {
+        Mapping incomingMapping = parseMapping(type, mappingSourceAsMap);
```

Comment:
This line doesn't need to be within the synchronized block I think?

Reply:
Yeah, it's misleading this way 🙂 
It is currently within the synchronized block and [I learned the hard way](https://github.com/elastic/elasticsearch/pull/97199#discussion_r1249541239) that it does need to be within it.